internal/controllers/utils/events.go

```diff
@@ -0,0 +1,116 @@
+type EventHistory struct {
+	History []*Event
+}
+
+// IsTimedOut proceses a process history an determines the overall state implied
+// by the events.
+// returns true if for any ObjectID:
+// - the overall state was InProgress from the initial time for more than timeout
+// - the overall state was InProgress from the last transition to InProgress for more than timeout
+// returns false if for any ObjectID:
+// - the last transition is Completed
+// - no events to process
+func (h EventHistory) IsTimedOut(now time.Time, timeout time.Duration) bool {
+	// Sort events chronologically from old to new
+	sort.Slice(h.History, func(i, j int) bool {
+		return h.History[i].Timestamp.Before(h.History[j].Timestamp)
+	})
+
+	// currentState stores the running value of the State (InProgress or Completed) for each ObjectID
+	currentState := map[string]string{}
+	// Initialize the current state to NonCompliant
+	for _, event := range h.History {
+		currentState[event.ObjectID] = InProgress
+	}
+	// initialTime records the initial Event for all ObjectIDs
+	initialTime := time.Time{}
+	// resetTime is a recording a zero time
+	resetTime := time.Time{}
+	// lastCompleted records the last time the overall state for all event transitioned to Completed
+	lastCompleted := time.Time{}
+	// lastInProgress records the last time the overall state for all event transitioned to InProgress
+	lastInProgress := time.Time{}
+	// Process each event chronologically one by one and recalculate the
+	// current state (InProgress or Completed)
+	for _, event := range h.History {
+		// If this is the first event for this ObjectID, record the initial time
+		if initialTime == resetTime {
+			initialTime = event.Timestamp
+		}
+		// If the state has not changed for this ObjectID, continue to the next event
+		if currentState[event.ObjectID] == event.State {
+			continue
+		}
+		// Update the current state for the ObjectID
+		currentState[event.ObjectID] = event.State
+		// Calculate the new overall state (InProgress or Completed) for all ObjectIDs
+		overallState := Completed
+		for _, state := range currentState {
+			if state != Completed {
+				overallState = InProgress
+				break
+			}
+		}
+		// Record the lastCompleted overall state
+		if overallState == Completed {
+			lastCompleted = event.Timestamp
+		}
+		// Record the lastCompleted overall state
+		if overallState == InProgress {
+			lastInProgress = event.Timestamp
+		}
+	}
+	if lastInProgress.After(lastCompleted) && lastCompleted != resetTime {
+		// Overall state is not Completed start timing out after the lastInProgress
+		return now.Sub(lastCompleted) > timeout
+	}
+	if lastCompleted.After(lastInProgress) && lastCompleted != resetTime {
+		// Overall state is Completed
+		return false
+	}
+	if lastCompleted == resetTime && initialTime != resetTime {
+		// Overall state is in progress from the beginning
+		return now.Sub(initialTime) > timeout
+	}
+	// If no Events are present, we can't timeout
+	return false
```

Comment:
It's very unlikely that all policies would have no status unless there's issue with the ACM. But, in theory, I think if enforced policies don't have a compliance status within the configured time, it should result in a timeout.

Reply:
If there is no event history, wouldn't it also mean that remediation hasn't started yet ? Before the cluster is ready I think to remember that there is not history or status.

Reply:
Yeah, I mean once the managed cluster is ready, policies should have a status shortly after, unless something is wrong. The cluster ready status is checked before the timeout, so if it reaches this point and there's no status for a long time, it's likely that something has gone wrong.